cpp_exercises/day2/date/date.cpp

```diff
@@ -137,3 +138,43 @@ void Date::fromJulianDay(int64_t julianDay) {
 void Date::addDays(int days) {
     fromJulianDay(toJulianDay() + days);
 }
+
+Date Date::fromString(const std::string& string) {
+    if(std::count(string.begin(), string.end(), ' ') != 0)
+        return Date();
+
+    if(std::count(string.begin(), string.end(), '-') != 2)
+        return Date();
+
+    std::string dayString, monthString, yearString;
+    Date::splitDayMonthYearString(string, dayString, monthString, yearString);
+
+    if(dayString.size() != 2 || monthString.size() != 2 || yearString.size() != 4)
+        return Date();
+
+    if(!Date::areStringValidInteger(dayString) || !Date::areStringValidInteger(monthString) || !Date::areStringValidInteger(yearString))
+        return Date();
+
+    int day = std::stoi(dayString);
+    int month = std::stoi(monthString);
+    int year = std::stoi(yearString);
+
+    if(!Date::isDateValid(day, month, year))
+        return Date();
+
+    return Date(day, month, year);
+}
+
+bool Date::areStringValidInteger(const std::string& string) {
+    char* p;
+    strtol(string.c_str(), &p, 10);
+    return *p == 0;
+}
+
+void Date::splitDayMonthYearString(const std::string& string, std::string& dayString, std::string& monthString, std::string& yearString) {
```

Comment:
Una buona pratica in generale quando si hanno degli out-parameter in una funzione, è quella di averli come puntatori anzichè reference.
Il motivo è semplice: l'utilizzo che ne fa il chiamante.
Il chiamante userà questo metodo (anche se privato) invocandolo come
```c++
splitDayMonthYearString(stringa_1, stringa_2, stringa_3, stringa_4);
```
ma visto così non è immediatamente chiaro quali parametri sono in sola lettura e quali in scrittura.
Avendo gli out-parameter specificati come puntatori invece è immediatamente chiaro che gli ultimi 3 saranno sovrascritti:
```c++
splitDayMonthYearString(stringa_1, &stringa_2, &stringa_3, &stringa_4);
```